api/envoy/api/v2/core/base.proto

```diff
@@ -95,6 +95,78 @@ message Metadata {
   map<string, google.protobuf.Struct> filter_metadata = 1;
 }
 
+// MetadataMatcher provides a general interface to check if a given value is matched in
+// :ref:`Metadata <envoy_api_msg_core.Metadata>`. It uses `filter` and `path` to retrieve the value
+// from the Metadata and then check if it's matched to one of the specified values.
+//
+// An example use of MetadataMatcher is specifying additional metadata in envoy.filters.http.rbac to
+// enforce access control based on dynamic metadata in a request.
+message MetadataMatcher {
```

Comment:
I have a general comment here about something that's been bothering me that I think we should get on top of now before it gets even worse.

We have a ton of different matching constructs in the API that at the end of the day are all basically the same. This includes and/or logic, headers, matching on various other things such as response flags, health check, etc.

I suggest we make a new namespace/folder called "matchers" and start moving all of the different matchers into that namespace. Also, I just looked, and is `StringMatch` used anywhere? I don't think so, so can we delete it as part of this change? 

Thoughts? cc @htuch

Reply:
I can start this change in this PR, (Initially only for MetadataMatcher and then move other Matchers to it in follow up PR).
As for `StringMatch`, I think it makes sense to make it a standalone Matcher under that new directory/namespace, at least it could be used by this MetadataMatcher.

Should I start it right in this PR? Thanks!
cc @lizan

Reply:
+1 I would like to see more standardization across the entire API of these match patterns.

Reply:
Added a new directory `api/envoy/type/matchers` and namespace `Envoy::Matchers` for the matchers. Currently it doesn't include and/or logic, headers, health check, etc. I prefer to do those in a different PR with a clear description of use case and goal.
Thanks!